src/main/java/cubicchunks/util/WatchersSortingList.java

```diff
@@ -0,0 +1,200 @@
+    private Object[] data = new Object[32768];
+    /**
+     * First used element of a data array. Leave some unused space to append
+     * elements with high priority.
+     */
+    private int start = data.length / 4;
+    /**
+     * Size of a list. Used to limit accessible members of data array during
+     * sorting and iteration.
+     */
+    private int size = 0;
+    /**
+     * Contain all data without certain order. Used to detect if element is
+     * already added
+     */
+    private final Set<T> dataAsSet = new HashSet<T>();
+    /** Contain elements which will be removed during tick event */
+    private final Set<T> toRemove = new HashSet<T>();
+
+    public WatchersSortingList(Comparator<T> orderIn) {
+        order = orderIn;
+    }
+
+    /** Sort elements */
+    public void sort() {
+        Arrays.sort((T[]) data, start, start + size, order);
+    }
+
+    /**
+     * Check if list is empty. Even when it return true list still could contain
+     * data, but it will be removed in attempt to access it via iterator.
+     * 
+     * @return if list contain any accessible data
+     */
+    public boolean isEmpty() {
+        return size - toRemove.size() == 0;
+    }
+
+    /**
+     * Return iterator over elements of list. If iterator meet any element which
+     * is scheduled to be removed by {@code remove(T element)} function, it will
+     * be removed from data.
+     * 
+     * @return iterator over elements
+     */
+    public Iterator<T> iterator() {
+        return new Iterator<T>() {
+
+            int a = start;
+            T prev = null;
+            T next = null;
+
+            private void peekNext() {
+                while (next == null && a < start + size) {
+                    T e = (T) data[++a - 1];
+                    if (toRemove.remove(e)) {
+                        dataAsSet.remove(e);
+                        data[--a] = data[start + --size];
+                    } else {
+                        next = e;
+                    }
+                }
+            }
+
+            @Override
+            public boolean hasNext() {
+                peekNext();
+                return next != null;
+            }
+
+            @Override
+            public T next() {
+                peekNext();
+                if (next == null)
+                    throw new ArrayIndexOutOfBoundsException();
+                prev = next;
+                next = null;
+                return prev;
+            }
+
+            @Override
+            public void remove() {
+                dataAsSet.remove(prev);
+                data[--a] = data[start + --size];
+            }
+        };
+    }
+
+    /**
+     * Schedule element to be removed. Function is not save, you need to check
+     * explicitly if this list contain element which need to be removed. Attempt
+     * to remove non-existing elements will cause nasty bugs.
+     */
+    public void remove(T entry) {
+        toRemove.add(entry);
```

Comment:
how does it handle removing element that doesn't exist in the list? I'm pretty sure it breaks size and isEmpty. If you don't want to handle such things, at least crash in those cases (PlayerCubeMap shouldn't be doing that). Same for other strange cases like adding the same thing twice.